### ISSUE TITLE
pgAdmin 4 connection problems on MacOS as non admin user

### ISSUE THREAD

**Issue by kelbstf** [pgadmin-org/pgadmin4#5579]
Hello,

i try to run 

- Postgres.app version 2.5.12 (171) together with 
- pgAdmin4 version 6.16 (4280.88) on 
- MacOS 10.15.7 (19H2026).

If i do so as administrative account (the one i also used to install both applications) on that MacOS,, everything works.
If i do so as non-administrative account, i can't register the PostgreSQL server in pgAdmin, apparently due to connection problems (timeout):

````
Unable to connect to server at "127.0.0.1", port 5432 failed: timeout expired
````

Both "127.0.0.1" and "localhost" will fail for the non administrative user.
But i can login to the same server using the psql client.

The log file for pgAdmin does not look different for both connections, at least if debug mode is not enabled...

I guess this might be a permissions problem.
Possibly filesystem permissions on a socket file or sth. similar, 
but since it does not work for TCP connections either, i currently have no idea where to look...

Does anyone have an idea?

**Comment by kelbstf** [comment]
Hello,
since i am really keen on using PostgreSQL, i just wanted to check back.
It would certainly help me already i someone could propose a specific (additional) log file where to start looking, and how to turn that on, if that would be a prerequisite.
Thanks

**Comment by yogeshmahajan-1903** [comment]
@kelbstf 
Can you please try below commands with user with which pgAdmin is installed - 
```
cd /Applications
sudo chmod -R 755 pgAdmin\ 4.app
```

With this , I was able to open pgAdmin with user other than who installed it. Also was able to connect to server.

**Comment by yogeshmahajan-1903** [comment]
Duplicate of #5613

**Comment by kelbstf** [comment]
@yogeshmahajan-1903 

actually i applied 777 in order to make the application launch (which is overdone), but worked:

````
 ls -al /Applications/pgAdmin\ 4.app/Contents/
total 216
drwxrwxrwx  11 root  wheel    352 Nov 17 12:28:39 2022 ./
drwxrwxrwx   3 root  wheel     96 Nov 17 12:23:18 2022 ../
-rwxrwxrwx   1 root  wheel  95436 Nov 17 12:28:18 2022 DEPENDENCIES*
drwxrwxrwx   8 root  wheel    256 Nov 17 12:29:30 2022 Frameworks/
-rwxrwxrwx   1 root  wheel   2089 Nov 17 12:24:45 2022 Info.plist*
-rwxrwxrwx   1 root  wheel   1026 Nov 17 12:28:18 2022 LICENSE*
drwxrwxrwx   3 root  wheel     96 Nov 17 12:29:31 2022 MacOS/
-rwxrwxrwx   1 root  wheel      9 Nov 17 12:24:45 2022 PkgInfo*
drwxrwxrwx  66 root  wheel   2112 Nov 17 12:27:41 2022 Resources/
drwxrwxrwx   6 root  wheel    192 Nov 17 12:28:40 2022 SharedSupport/
drwxrwxrwx   3 root  wheel     96 Nov 17 12:28:39 2022 _CodeSignature/
````

The non administrative user also has access:

````
ls -al /Applications/pgAdmin\ 4.app/Contents/Resources/web/pgAdmin4.py
-rwxrwxrwx  1 root  wheel  8682 Nov 17 12:28:13 2022 /Applications/pgAdmin 4.app/Contents/Resources/web/pgAdmin4.py*
````


````
cat /Applications/pgAdmin\ 4.app/Contents/Resources/web/pgAdmin4.py
##########################################################################
#
# pgAdmin 4 - PostgreSQL Tools
#
# Copyright (C) 2013 - 2022, The pgAdmin Development Team
# This software is released under the PostgreSQL Licence
#
##########################################################################
[ ... ]
````

But everything related to running or creating the server then fails.

In the application log i find complaints about the websocket and SSL certificate problems.
The latter is because of a upstream firewall which enforces SSL scanning, and the python bundled with pgAdmin is apparently not picking up the firewall's CA certificate from the MacOS certificate store.
I deem that unrelated to the actuall problem, as that SSL problem also occurs for the administrative MacOS user.

Therefore, might it be possible that the PG serverwants to create a websocket where it can't because it is running in the context of the nonadministrative user?

I tried both with "localhost" and "127.0.0.1"...


````
pgAdmin Runtime Environment
--------------------------------------------------------
Python Path: "/Applications/pgAdmin 4.app/Contents/Frameworks/Python.framework/Versions/Current/bin/python3"
Runtime Config File: "/Users/nonadminuser/Library/Preferences/pgadmin/runtime_config.json"
pgAdmin Config File: "/Applications/pgAdmin 4.app/Contents/Resources/web/config.py"
Webapp Path: "/Applications/pgAdmin 4.app/Contents/Resources/web/pgAdmin4.py"
pgAdmin Command: "/Applications/pgAdmin 4.app/Contents/Frameworks/Python.framework/Versions/Current/bin/python3 -s /Applications/pgAdmin 4.app/Contents/Resources/web/pgAdmin4.py"


[ ... ]


Total spawn time to start the pgAdmin4 server: 0.045 Sec
2022-12-15 13:40:53,815: WARNING	werkzeug:	WebSocket transport not available. Install simple-websocket for improved performance.

 * Serving Flask app 'pgadmin' (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off

------------------------------------------
Total time taken to ping pgAdmin4 server: 20.374 Sec
------------------------------------------
Total launch time of pgAdmin4: 20.54 Sec
------------------------------------------
Application Server URL: http://127.0.0.1:49712/?key=88f68700-227d-42a8-b9e1-1c0158980cbd
2022-12-15 13:40:55,478: ERROR	pgadmin:	Exception when checking for update
Traceback (most recent call last):
  File "/Applications/pgAdmin 4.app/Contents/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 1346, in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
  File "/Applications/pgAdmin 4.app/Contents/Frameworks/Python.framework/Versions/3.9/lib/python3.9/http/client.py", line 1285, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/Applications/pgAdmin 4.app/Contents/Frameworks/Python.framework/Versions/3.9/lib/python3.9/http/client.py", line 1331, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/Applications/pgAdmin 4.app/Contents/Frameworks/Python.framework/Versions/3.9/lib/python3.9/http/client.py", line 1280, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/Applications/pgAdmin 4.app/Contents/Frameworks/Python.framework/Versions/3.9/lib/python3.9/http/client.py", line 1040, in _send_output
    self.send(msg)
  File "/Applications/pgAdmin 4.app/Contents/Frameworks/Python.framework/Versions/3.9/lib/python3.9/http/client.py", line 980, in send
    self.connect()
  File "/Applications/pgAdmin 4.app/Contents/Frameworks/Python.framework/Versions/3.9/lib/python3.9/http/client.py", line 1454, in connect
    self.sock = self._context.wrap_socket(self.sock,
  File "/Applications/pgAdmin 4.app/Contents/Frameworks/Python.framework/Versions/3.9/lib/python3.9/ssl.py", line 500, in wrap_socket
    return self.sslsocket_class._create(
  File "/Applications/pgAdmin 4.app/Contents/Frameworks/Python.framework/Versions/3.9/lib/python3.9/ssl.py", line 1040, in _create
    self.do_handshake()
  File "/Applications/pgAdmin 4.app/Contents/Frameworks/Python.framework/Versions/3.9/lib/python3.9/ssl.py", line 1309, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1129)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Applications/pgAdmin 4.app/Contents/Resources/web/pgadmin/browser/__init__.py", line 487, in check_browser_upgrade
    response = urlopen(url, data, 5, cafile=config.CA_FILE)
  File "/Applications/pgAdmin 4.app/Contents/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 214, in urlopen
    return opener.open(url, data, timeout)
  File "/Applications/pgAdmin 4.app/Contents/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 517, in open
    response = self._open(req, data)
  File "/Applications/pgAdmin 4.app/Contents/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 534, in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
  File "/Applications/pgAdmin 4.app/Contents/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 494, in _call_chain
    result = func(*args)
  File "/Applications/pgAdmin 4.app/Contents/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 1389, in https_open
    return self.do_open(http.client.HTTPSConnection, req,
  File "/Applications/pgAdmin 4.app/Contents/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 1349, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1129)>
````

**Comment by yogeshmahajan-1903** [comment]
@kelbstf 
Above error is related to checking update for pgAdmin version. I do not see any logs indicating connection to server is failed.
Can you please provide screen shot where you are able to connect using psql with non-admin user?
Also, a pgAdmin error screenshot?

**Comment by kelbstf** [comment]
@yogeshmahajan-1903

that was a bit tough. 
I believe this was an absolute niche issue.

The problem in my situation was that when creating a new server, pgAdmin wants to access the user's keychain, no matter if the user also antered a password for the server. 
Because the keychain wasn't accessible due to Keychain.app internal problems (that for odd reasons did not manifest for other applications and therefore remained uncovered so far), that access failed silently in the background, and pgAdmin just notified that connection to the server 127.0.0.1 was refused.

After having fixed the keychain problem, everything works as expected. 
The only customisation required for the current version 6.17 of pgAdmin was the permission scheme to be changed to 755 on the application's folder, as you mentioned above .

Thank you very much for your help.


Best

**Comment by yogeshmahajan-1903** [comment]
Closing issue as it fixed with #5613
